wallet/db.c

```diff
@@ -858,6 +858,15 @@ static struct migration dbmigrations[] = {
 
     /* Issue #4887: reset the payments.id sequence after the migration above. Since this is a SELECT statement that would otherwise fail, make it an INSERT into the `vars` table.*/
     {SQL("/*PSQL*/INSERT INTO vars (name, intval) VALUES ('payment_id_reset', setval(pg_get_serial_sequence('payments', 'id'), COALESCE((SELECT MAX(id)+1 FROM payments), 1)))"), NULL},
+
+    /* Issue #4901: Partial index speeds up startup on nodes with ~1000 channels.  */
+    {SQL("CREATE INDEX channel_htlcs_speedup_unresolved_idx"
+	 "    ON channel_htlcs(channel_id, direction)"
+	 " WHERE hstate NOT IN (9, 19);"),
+	 /*  9 = RCVD_REMOVE_ACK_REVOCATION
+	  * 19 = SENT_REMOVE_ACK_REVOCATION
```

Comment:
It would be really nice if we could put in the constants here instead of hardcoding the constant values 9 and 19.  Any ideas how to do this?  Could use a function for this migration so we can `db_bind_int`, but in principle a function is not necessary and would be too heavyweight...

Reply:
maybe it is something stupid and I'm missing the point, but we could define a `#define RCVD_REMOVE_ACK_REVOCATION 9` somewhere?

Just worry if it is a valid type of `db_bind_int`

Reply:
They are already defined here: https://github.com/ElementsProject/lightning/blob/a9e261fc20b22d5a77a55276be442ede5baf4fef/common/htlc_state.h#L22 https://github.com/ElementsProject/lightning/blob/a9e261fc20b22d5a77a55276be442ede5baf4fef/common/htlc_state.h#L36

The `#define` would duplicate the above declaration, and if not carefully ordered relative to the above, would cause a syntax error.  We also prefer to use the above `enum` declaration if possible.

Reply:
Rather than asserting the values of these constants in a comment, you could use a compile-time assertion. I see you have a `BUILD_ASSERT` macro presumably for that purpose. `wallet/wallet.h` does contain assertions of the numeric values of these enumerators, but you could duplicate those assertions here in `wallet/db.c` at zero cost to make absolutely certain that the enumerators retain the values you're expecting.

Reply:
It is not really about *asserting* the values, it is more that stylistically, I would prefer not to see hardcoded numbers in code.  Ideally I would be able to do something like `tal_fmt("CREATE INDEX blah blah WHERE hstate NOT IN (%d, %d);;", RCVD_REMOVE_ACK_REVOCATION, SENT_REMOVE_ACK_REVOCATION)` but that cannot work as this is a static data structure and this is C, sigh.  C++ initializers...

I think `BUILD_ASSERT` would have to be outside the `db_migrations` array, too.  And the point is not to check the numbers (it is already done elsewhere, as you note) but as a reminder to future readers of the code what these hardcoded `9` and `19` are.  If `BUILD_ASSERT` has to be done outside of `db_migrations`, then as `db_migrations` grows, the explanation of the hardcoded `9` and `19` becomes more distant from the query.  In that respect a comment is superior as it can sit right next to that query in perpetuam (and the check is already done elsewhere).

Reply:
Then you should at least put a comment adjacent to the assertions in `wallet/wallet.h` to help future coders understand what they will be breaking if they violate the assertions.

Are you certain you can't put an assertion inside your `const char *[]` initializer? What about using a compound statement expression for the initializer value? `({ BUILD_ASSERT(…); "string constant here"; })`

Reply:
Does not work, not allowed in initializers.  Like I said, C++ initializers allow this, but not C, because C++ actually hooks an extra function into `main` that calls all initializers, but this it not available in C.

```c
/*example.c*/
#define BUILD_ASSERT(cond) \
        do { (void) sizeof(char [1 - 2*!(cond)]); } while(0)

struct example {
        const char *string;
};

struct example db_migrations[] = {
        {({BUILD_ASSERT(1); "example"; })},
};
```

```
$ gcc example.c
example.c:10:3: error: braced-group within expression allowed only inside a function
   10 |  {({BUILD_ASSERT(1); "example"; })},
      |   ^
```

Reply:
> Then you should at least put a comment adjacent to the assertions in `wallet/wallet.h` to help future coders understand what they will be breaking if they violate the assertions.

We already have a warning comment in `common/htlc_state.h` for this, which is more important since this is where the constants are actually defined: https://github.com/ElementsProject/lightning/blob/a9e261fc20b22d5a77a55276be442ede5baf4fef/common/htlc_state.h#L5-L8

Reply:
`BUILD_ASSERT_OR_ZERO` works:

```c
/* example.c */
#include<stddef.h>

#define BUILD_ASSERT_OR_ZERO(cond) \
        (sizeof(char [1 - 2*!(cond)]) - 1)

enum example {
        EXAMPLE_ZERO,
        EXAMPLE_ONE
};

struct db_entry {
        const char *query;
        const char *foo;
};

struct db_entry db_migrations[] = {
        { "string"
        + BUILD_ASSERT_OR_ZERO(EXAMPLE_ZERO == 0)
        + BUILD_ASSERT_OR_ZERO(EXAMPLE_ONE == 1),
          NULL
        },
};

int main() {
        return 0;
}
```

```
$ gcc -c example.c
```

Still suboptimal though since we do not really want to *assert*, we want to not have to use hardcoded numbers in the source code...

Reply:
Except clang hates it...

```
wallet/db.c:868:2: error: adding 'unsigned long' to a string does not append to the string [-Werror,-Wstring-plus-int]
        + BUILD_ASSERT_OR_ZERO( 9 == RCVD_REMOVE_ACK_REVOCATION)
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wallet/db.c:868:2: note: use array indexing to silence this warning
        + BUILD_ASSERT_OR_ZERO( 9 == RCVD_REMOVE_ACK_REVOCATION)
        ^
        [                                                       ]
```

bleah.....

Reply:
What about `&"string"[BUILD_ASSERT_OR_ZERO(…)]`?

Reply:
Current version already does that.